File: test-utils/src/main/java/ru/virgil/test_utils/TestUtils.java
package ru.virgil.test_utils;


import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.type.CollectionType;
import com.fasterxml.jackson.databind.type.TypeFactory;
import lombok.RequiredArgsConstructor;
import lombok.extern.apachecommons.CommonsLog;
import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;
import org.springframework.stereotype.Component;
import org.springframework.test.web.servlet.MvcResult;

import java.io.IOException;
import java.io.UnsupportedEncodingException;
import java.util.Collection;
import java.util.Map;
import java.util.Optional;
import java.util.stream.Collectors;

@Component
@RequiredArgsConstructor
@CommonsLog
public class TestUtils {

    // todo: оставить тут только работу с парсингом
    public static final int JSON_OUTPUT_LENGTH = 2048;
    public static final String TEST_RESULTS_TEMPLATE = """
                     
                        
            /// TEST RESULTS ///
                        
            REQUEST: %s %s%s -> %d
                        
            REQUEST JSON:
            %s
                        
            RESPONSE JSON:
            %s
            """;
    protected final ObjectMapper objectMapper;

    // todo: изввлечение коллекций?
    public <D> D extractDtoFromResponse(MvcResult mvcResult, Class<D> dtoClass) throws IOException {
        return objectMapper.readValue(mvcResult.getResponse().getContentAsString(), dtoClass);
    }

    @SuppressWarnings("rawtypes")
    public <D, C extends Collection, R> R extractDtoFromResponse(MvcResult mvcResult, Class<C> collectionClass,
            Class<D> dtoClass) throws IOException {
        TypeFactory typeFactory = objectMapper.getTypeFactory();
        CollectionType collectionType = typeFactory.constructCollectionType(collectionClass, dtoClass);
        String contentAsString = mvcResult.getResponse().getContentAsString();
        return objectMapper.readValue(contentAsString, collectionType);
    }

    public void printResponse(MvcResult mvcResult) {
        try {
            String method = mvcResult.getRequest().getMethod();
            String requestURI = mvcResult.getRequest().getRequestURI();
            String requestParams = extractPrettyParams(mvcResult);
            int status = mvcResult.getResponse().getStatus();
            String responseContent = extractPrettyResponse(mvcResult);
            String requestContent = Optional.ofNullable(mvcResult.getRequest().getContentAsString())
                    .map(this::extractPrettyRequest).orElse("NONE");
            log.info(TEST_RESULTS_TEMPLATE.formatted(method, requestURI, requestParams, status, requestContent,
                    responseContent));
        } catch (UnsupportedEncodingException | JsonProcessingException e) {
            throw new RuntimeException(e);
        }
    }

    protected String extractPrettyRequest(String requestContent) {
        try {
            requestContent = objectMapper.readTree(requestContent).toPrettyString();
            requestContent = shortenJson(requestContent);
            return requestContent;
        } catch (JsonProcessingException e) {
            throw new RuntimeException(e);
        }
    }

    protected String extractPrettyResponse(MvcResult mvcResult)
            throws UnsupportedEncodingException, JsonProcessingException {
        String responseContent = mvcResult.getResponse().getContentAsString();
        if (!isJson(responseContent)) {
            return "BODY: content-type -> " + mvcResult.getResponse().getContentType();
        }
        responseContent = objectMapper.readTree(responseContent).toPrettyString();
        responseContent = shortenJson(responseContent);
        if (responseContent.isEmpty()) {
            return "NONE";
        }
        return responseContent;
    }

    public boolean isJson(String jsonInString) {
        try {
            new JSONObject(jsonInString);
        } catch (JSONException ex) {
            try {
                new JSONArray(jsonInString);
            } catch (JSONException ex1) {
                return false;
            }
        }
        return true;
    }

    protected String shortenJson(String requestContent) {
        String substring = requestContent.substring(0, Math.min(JSON_OUTPUT_LENGTH, requestContent.length()));
        if (substring.length() == JSON_OUTPUT_LENGTH) {
            substring = substring + """
                    ...
                                        
                    /// JSON SHORTENED TO %d SYMBOLS ///
                                        
                    """.formatted(JSON_OUTPUT_LENGTH);
        }
        return substring;
    }

    protected String extractPrettyParams(MvcResult mvcResult) {
        Map<String, String[]> parameterMap = mvcResult.getRequest().getParameterMap();
        String stringParams = parameterMap.keySet().stream()
                .map(key -> key + "=" + String.join(",", parameterMap.get(key)))
                .collect(Collectors.joining("&", "?", ""));
        if (stringParams.equals("?")) {
            stringParams = "";
        }
        return stringParams;
    }

}

File: test-utils/src/main/java/ru/virgil/test_utils/TestUtils.kt
package ru.virgil.test_utils

import com.fasterxml.jackson.databind.ObjectMapper
import org.json.JSONArray
import org.json.JSONException
import org.json.JSONObject
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Component
import org.springframework.test.web.servlet.MvcResult
import java.util.*
import java.util.stream.Collectors
import kotlin.math.min

private const val ERROR_VALUE = "ERROR"
private const val NONE_VALUE = "NONE"
private const val JSON_OUTPUT_LENGTH = 2048

@Component
class TestUtils(protected val objectMapper: ObjectMapper) {

    private val logger = LoggerFactory.getLogger(this.javaClass)

    fun printResponse(mvcResult: MvcResult) {
        val method = mvcResult.request.method ?: ERROR_VALUE
        val requestURI = mvcResult.request.requestURI ?: ERROR_VALUE
        val requestParams = extractPrettyParams(mvcResult)
        val status = mvcResult.response.status
        val requestContent = extractPrettyRequest(mvcResult)
        val responseContent = extractPrettyResponse(mvcResult)
        logger.info(buildTestResults(method, requestURI, requestParams, status, requestContent, responseContent))
    }

    private fun extractPrettyRequest(mvcResult: MvcResult): String =
        Optional.ofNullable(mvcResult.request.contentAsString)
            .map { makeJsonPretty(it) }
            .orElse(NONE_VALUE)

    protected fun makeJsonPretty(requestContent: String): String {
        var prettyRequestContent = objectMapper.readTree(requestContent).toPrettyString()
        prettyRequestContent = shortenJson(prettyRequestContent)
        return prettyRequestContent
    }

    protected fun extractPrettyResponse(mvcResult: MvcResult): String {
        var responseContent = mvcResult.response.contentAsString
        return when {
            !responseContent.isJson() -> "BODY: content-type -> ${mvcResult.response.contentType}"
            else -> {
                responseContent = objectMapper.readTree(responseContent).toPrettyString()
                responseContent = shortenJson(responseContent)
                responseContent.ifEmpty { NONE_VALUE }
            }
        }
    }

    protected fun shortenJson(requestContent: String): String {
        var shortedRequestContent = requestContent.substring(0, min(JSON_OUTPUT_LENGTH, requestContent.length))
        if (shortedRequestContent.length == JSON_OUTPUT_LENGTH) {
            shortedRequestContent += buildShortenMessage()
        }
        return shortedRequestContent
    }

    protected fun extractPrettyParams(mvcResult: MvcResult): String {
        val parameterMap = mvcResult.request.parameterMap
        var stringParams = parameterMap.keys.stream()
            .map { "$it=${java.lang.String.join(",", *parameterMap[it])}" }
            .collect(Collectors.joining("&", "?", ""))
        if (stringParams == "?") {
            stringParams = ""
        }
        return stringParams
    }

}

private fun buildShortenMessage() = """
...
                    
/// JSON SHORTENED TO $JSON_OUTPUT_LENGTH SYMBOLS ///
                                      
"""

private fun buildTestResults(
    method: String, requestURI: String, requestParams: String, status: Int, requestContent: String,
    responseContent: String,
) = """
                                                    
/// TEST RESULTS ///
            
REQUEST: 
$method $requestURI$requestParams -> $status
            
REQUEST JSON: 
$requestContent
            
RESPONSE JSON: 
$responseContent

"""

private fun String.isJson(): Boolean {
    try {
        JSONObject(this)
    } catch (ex: JSONException) {
        try {
            JSONArray(this)
        } catch (ex1: JSONException) {
            return false
        }
    }
    return true
}
